kernel/multicore_bringup/src/aarch64.rs

```diff
@@ -0,0 +1,214 @@
+    ap_data.ap_data_virt_addr.write(VirtualAddress::new_canonical(&ap_data as *const _ as usize));
+    ap_data.ap_stage_two_virt_addr.write(VirtualAddress::new_canonical(ap_stage_two as usize));
+
+    let entry_point_phys_addr: PhysicalAddress;
+    let ap_data_phys_addr: PhysicalAddress;
+    let mut ap_startup_mapped_pages: MappedPages; // must be held throughout APs being booted up
+    {
+        let mut kernel_mmi = kernel_mmi_ref.lock();
+        let page_table = &mut kernel_mmi.page_table;
+
+        // get the physical address of ap_entry_point
+        let entry_point_virt_addr = VirtualAddress::new_canonical(ap_entry_point as usize);
+        entry_point_phys_addr = page_table.translate(entry_point_virt_addr).unwrap();
+
+        // get the physical address of the MmuConfig
+        let mmu_config_virt_addr = VirtualAddress::new_canonical(&mmu_config as *const _ as usize);
+        ap_data.ap_mmu_config.write(page_table.translate(mmu_config_virt_addr).unwrap());
+
+        // Note: this could be variable on aarch64
+        // we'd need a way to find any identity mappable page+frame range
+        const AP_STARTUP: usize = 0x10000;
+
+        // When the AArch64 core will enter startup code, it will do so with MMU disabled,
+        // which means these frames MUST be identity mapped.
+        let ap_startup_frames = memory::allocate_frames_by_bytes_at(PhysicalAddress::new_canonical(AP_STARTUP), PAGE_SIZE)
+            .map_err(|_e| "handle_ap_cores(): failed to allocate AP startup frames")?;
+        let ap_startup_pages  = memory::allocate_pages_at(VirtualAddress::new_canonical(AP_STARTUP), ap_startup_frames.size_in_frames())
+            .map_err(|_e| "handle_ap_cores(): failed to allocate AP startup pages")?;
+
+        // map this RWX
+        let flags = PteFlags::new().valid(true).writable(true);
+        ap_startup_mapped_pages = page_table.map_allocated_pages_to(
+            ap_startup_pages,
+            ap_startup_frames,
+            flags,
+        )?;
+
+        // get physical address of the ApTrampolineData structure
+        ap_data_phys_addr = page_table.translate(ap_data.ap_data_virt_addr.read()).unwrap();
+
+        // copy the entry point code
+        let dst = ap_startup_mapped_pages.as_slice_mut(0, PAGE_SIZE).unwrap();
+        let src = unsafe { (ap_entry_point as *const [u8; PAGE_SIZE]).as_ref() }.unwrap();
+        dst.copy_from_slice(src);
+    }
+
+    // Create a stack
+    let mut ap_stack = Some(NoDrop::new(stack::alloc_stack(
+        KERNEL_STACK_SIZE_IN_PAGES,
+        &mut kernel_mmi_ref.lock().page_table,
+    ).ok_or("could not allocate AP stack!")?));
+
+    log::error!("Bruteforcing 256 CpuIds that are likely to be valid");
+    for aff1 in 0..16 {
+        for aff0 in 0..16 {
+            let cpu_id = CpuId::new(0, 0, aff1, aff0);
+
+            // Write the per-core fields of ApTrampolineData
+            let stack = ap_stack.take().unwrap();
```

Comment:
i love the idea of reusing stacks for CPUs that don't actually exist. if you reorganize this loop a bit, you can easily avoid some complexity and also the leaking of the last stack.

Reply:
Good point. I did what you advised.